expose-daemon.ps1

```diff
@@ -0,0 +1,149 @@
+          + [System.Convert]::ToBase64String($privateKeyFromCert.Key.Export([System.Security.Cryptography.CngKeyBlobFormat]::Pkcs8PrivateBlob), [System.Base64FormattingOptions]::InsertLineBreaks) `
+          + "`n-----END PRIVATE KEY-----");
+    Encoding = "Ascii";
+  }
+  Set-Content @parms
+
+  Write-Host "`n=== Generating Client certificate"
+  $parms = @{
+    CertStoreLocation = "Cert:\LocalMachine\My";
+    Subject = "CN=clientCert";
+    Signer = $rootCert ;
+    KeyExportPolicy = "Exportable";
+    Provider = "Microsoft Enhanced Cryptographic Provider v1.0";
+    TextExtension = @("2.5.29.37= {text}1.3.6.1.5.5.7.3.2") ;
+    HashAlgorithm = "sha256";
+    KeyLength = 4096;
+  }
+  $clientCert = New-SelfSignedCertificate  @parms
+
+  $parms = @{
+    Path = "$clientCertsPath\cert.pem" ;
+    Value = ("-----BEGIN CERTIFICATE-----`n" + [System.Convert]::ToBase64String($clientCert.RawData, [System.Base64FormattingOptions]::InsertLineBreaks) + "`n-----END CERTIFICATE-----");
+    Encoding = "Ascii";
+  }
+  Set-Content @parms
+
+  Write-Host "`n=== Generating Client key"
+  $clientprivateKeyFromCert = [System.Security.Cryptography.X509Certificates.RSACertificateExtensions]::GetRSAPrivateKey($clientCert)
+  $parms = @{
+    Path = "$clientCertsPath\key.pem";
+    Value = ("-----BEGIN PRIVATE KEY-----`n" `
+          + [System.Convert]::ToBase64String($clientprivateKeyFromCert.Key.Export([System.Security.Cryptography.CngKeyBlobFormat]::Pkcs8PrivateBlob), [System.Base64FormattingOptions]::InsertLineBreaks) `
+          + "`n-----END PRIVATE KEY-----");
+    Encoding = "Ascii";
+  }
+  Set-Content @parms
+
+  copy $serverCertsPath\ca.pem $clientCertsPath\ca.pem
+}
+
+function updateConfig($daemonJson, $serverCertsPath) {
+  $config = @{}
+  if (Test-Path $daemonJson) {
+    $config = (Get-Content $daemonJson) -join "`n" | ConvertFrom-Json
+  }
+
+  $config = $config | Add-Member(@{ `
+    hosts = @("tcp://0.0.0.0:2376", "npipe://"); `
+    tlsverify = $true; `
+    tlscacert = "$serverCertsPath\ca.pem"; `
+    tlscert = "$serverCertsPath\server-cert.pem"; `
+    tlskey = "$serverCertsPath\server-key.pem" `
+    }) -Force -PassThru
+
+  Write-Host "`n=== Creating / Updating $daemonJson"
+  $config | ConvertTo-Json | Set-Content $daemonJson -Encoding Ascii
+}
+
+$dockerData = "$env:ProgramData\docker"
+$userPath = "$env:USERPROFILE\.docker"
+
+ensureDirs @("$dockerData\certs.d", "$dockerData\config", "$userPath")
+
+$serverCertsPath = "$dockerData\certs.d"
+$clientCertsPath = "$userPath"
+$rootCert = createCA "$dockerData\certs.d"
+
+createCerts $rootCert $serverCertsPath $serverName $ipAddresses $clientCertsPath
+updateConfig "$dockerData\config\daemon.json" $serverCertsPath $enableLCOW $experimental
+
+Write-Host "Restarting Docker"
+Restart-Service docker
+
+Write-Host "Opening Docker TLS port"
+& netsh advfirewall firewall add rule name="Docker TLS" dir=in action=allow protocol=TCP localport=2376
```

Comment:
It might be nice to add debug statements here so that it's obvious what went wrong if something fails to run.

Reply:
@BradleyBoutcher can this be added to one of your follow-up issues too?

Reply:
`Write-Host` is the equivalent of `echo`. I think the script gives a fair bit of information of how it is proceeding so that you know where you are when it fails, then the command itself will provide the actual error